### PR TITLE
RenameUpload implementation with flysystem

### PR DESCRIPTION
**Pull request by svycka**
What do you think about this?
I am not sure about fallback to local adapter.

### CODE REVIEW

File: src/Filter/File/RenameUpload.php [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+     * @return FilesystemInterface
+     */
+    public function getFilesystem()
+    {
+        if (!$this->filesystem) {
+            // Fallback to local and guess required root path.
+            $root = $this->guessRootPath();
+
+            // Warning: changing target path later will not update root path.
+            $this->filesystem = new Filesystem(new LocalAdapter($root));
```
Comment (svycka) [inline src/Filter/File/RenameUpload.php]: maybe this is not good idea. throw exception instead?

**Comment by basz** [comment]
Hmm me neither. I think it is ok if the filter just fails without an filesystem set.

I'm wondering if it would be useful if the flysystem pluginmanager would internally be available as a service (via factory). 

The user could then specify a flysystem name in the filter configuration instead of a a concrete implementation.

factory would put from sm
$serviceLocator->get('BsbFlysystemManager');


function __constuct(FlysystemManager $flysystemManager) {
$this->flysystemManager = flysystemManager
}
function setFilesystem($fileystem) {
if (is_string($filestem)) {
 $filesstem = $this->pluginManager->get($fileystem);
}


}

i got to run outside, but -if you understand me- what do you think?

File: src/Filter/File/RenameUpload.php (hunk 1 above)
Comment (basz) [inline src/Filter/File/RenameUpload.php]: agreed, see below

**Comment by svycka** [comment]
I don't like the idea having manager as dependency.
instead of this:
```php
$filter = new \BsbFlysystem\Filter\File\RenameUpload([
    'target' => 'path/to/file.txt',
    'filesystem' => $serviceLocator->get('BsbFlysystemManager')->get('awss3')
]);
```
you would have this:
```php
$filter = new \BsbFlysystem\Filter\File\RenameUpload($serviceLocator->get('BsbFlysystemManager'), [
    'target' => 'path/to/file.txt',
    'filesystem' => 'awss3'
]);
```
Yes this would be good when added to config in form field but still i don't want it.

Also you can create factories for \BsbFlysystem\Filter\File\RenameUpload this would eliminate BsbFlysystemManager or adapters as dependencies at all.

But if you think this is must have feature I will add it.

File: src/Filter/File/RenameUpload.php [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    protected $filesystem;
+
+    /**
+     * @throws UnexpectedValueException
+     * @return FilesystemInterface
+     */
+    public function getFilesystem()
+    {
+        if (!$this->filesystem) {
+            throw new UnexpectedValueException('Missing required filesystem.');
```
Comment (svycka) [inline src/Filter/File/RenameUpload.php]: Maybe throw exception in constructor? but I think this better.

**Comment by basz** [comment]
I think we should start as simple as possible so i'm tempted to forget about injecting dependancies for now. However since this cannot be changed later on without BC what do you think about supporting both signatures.

```
$options['filesystem'] // optional if filesystem_by_name is filesystem
$options['filesystem_by_name']; // optional if filesystem is used filesystem_by_name
$options['filesystem_manager']; // required is filesystem_by_name
__construct($options) {
      if (null !== $options) {
          $this->setOptions($options);
      }
}

setFilesystem(Filesystem $filesystem)
setFilesystemByName($name) {
   $this->filesystem = null;
   $this->filesystem_by_name = $name;
}
setFilesystemManager(BsbFlysystemManager $fsm)

getFilesystemManager() {
  if (!$this->filesystem_manager) {
      throw new RuntimeException('Boo');
   }
   return $this->filesystem_manager;
}

__invoke() {
   if ($this->filesystem === null && $this->filesystem_by_name !== null) {
      $this->filesystem = this->getFilesystemManager()->get($this->filesystem_by_name);
   }

   // ... 
}

```

A factory could be provided and registered to the filter manager which injects the manager only (filter manager should have the shared flag set to false).

Best of both worlds?
```

```

**Comment by svycka** [comment]
What did you mean this can't be changed without BC? If we leave for now as it is we can change later to your suggested behavior without BC but not the other way around.

I still don't like manager as dependency.

Your form will have \BsbFlysystem\Filter\File\RenameUpload as dependancy so how it is configured does not matter and \BsbFlysystem\Filter\File\RenameUpload does not need manager.

You always can create factories for RenameUploadS3, RenameUploadLocal... and add them in form config.

I would say we go for now without manager and later we can add it like you suggested without any BC if really needed.

File: test/Filter/File/RenameUploadTest.php [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+    public function setUp()
+    {
+        $this->filesystem = $this->prophesize('League\Flysystem\FilesystemInterface');
+    }
+
+    public function testCanUploadFile()
+    {
+        $path = 'path/to/file.txt';
+        $this->filesystem->putStream($path, Argument::any())
```
Comment (svycka) [inline test/Filter/File/RenameUploadTest.php]: not sure about Argument::any() maybe better would be to at least check that type is resource? Unfortunately don't know how.

**Comment by svycka** [comment]
Also maybe add zend-filter to composer.json just not sure as suggested or required

**Comment by basz** [comment]
re: zend-filter to composer.json
required for dev (due to tests)
suggested for non dev

re: BC. off course you are right, $options can contain anything. First glance I thought the constructor would take concrete instances as arguments, but none of the zf2 filters have that. So yes, forget about the manager for now.

**Comment by svycka** [comment]
I think it's ready to be merged.

**Comment by basz** [comment]
thank you, i'll prep a 1.3.0 release soon

**Pull request merged**